cpp/ql/test/experimental/query-tests/Security/CWE/CWE-416/test.cpp

```diff
@@ -0,0 +1,719 @@
+
+typedef unsigned long size_t;
+
+template<class T>
+struct remove_const { typedef T type; };
+
+template<class T>
+struct remove_const<const T> { typedef T type; };
+
+// `remove_const_t<T>` removes any `const` specifier from `T`
+template<class T>
+using remove_const_t = typename remove_const<T>::type;
+
+template<class T>
+struct remove_reference { typedef T type; };
+
+template<class T>
+struct remove_reference<T &> { typedef T type; };
+
+template<class T>
+struct remove_reference<T &&> { typedef T type; };
+
+// `remove_reference_t<T>` removes any `&` from `T`
+template<class T>
+using remove_reference_t = typename remove_reference<T>::type;
+
+template<class T>
+struct decay_impl {
+    typedef T type;
+};
+
+template<class T, size_t t_size>
+struct decay_impl<T[t_size]> {
+    typedef T* type;
+};
+
+template<class T>
+using decay_t = typename decay_impl<remove_reference_t<T>>::type;
+
+
+namespace std
+{
+	template<class T> constexpr T&& forward(remove_reference_t<T>& t) noexcept;
+	template<class T> constexpr T&& forward(remove_reference_t<T>&& t) noexcept;
+}
+
+// --- iterator ---
```

Comment:
Is there anything stopping us moving some of these headers into common files (shared across all tests that need them)?  It's something that's come up before, but the particularly large number of `std` headers in this test is making me think this improvement might be due.

Reply:
We should totally work towards that, yes. I'm not quite sure how to do this since each of our tests live in separate directories with no common root so, as far as I know, we can't just `#include "some-directory.h"`

Reply:
Once upon a time we could include files outside the test directory, then we couldn't, at this point I genuinely don't know.

Reply:
I've written this up as an issue, since it was bothering me.

Reply:
Thanks!